python/utils/train_model.py

```diff
@@ -0,0 +1,313 @@
+        labels.extend([label] * n_subwords)
+
+    return tokenized_sentence, labels
+
+def flat_accuracy(preds, labels):
+
+    """
+    Define a flat accuracy metric to use while training the model.
+    """
+
+    pred_flat = np.argmax(preds, axis=2).flatten()
+    labels_flat = labels.flatten()
+
+    return np.sum(pred_flat == labels_flat) / len(labels_flat)
+
+###
+
+# Read training and dev data in using SentenceGetter
+getter_train = SentenceGetter(train_data_path)
+getter_dev = SentenceGetter(dev_data_path)
+
+# Prepare and convert the data to appropriate formats
+train, dev = BertDataset(getter_train), BertDataset(getter_dev)
+
+# Input IDs (tokens), tags (label IDs), attention masks
+tr_inputs = torch.tensor(train.input_ids)
+val_inputs = torch.tensor(dev.input_ids)
+tr_tags = torch.tensor(train.tags)
+val_tags = torch.tensor(dev.tags)
+tr_masks = torch.tensor(train.attn_masks)
+val_masks = torch.tensor(dev.attn_masks)
+
+train_data = TensorDataset(tr_inputs, tr_masks, tr_tags)
+train_sampler = RandomSampler(train_data)
+train_dataloader = DataLoader(train_data, sampler=train_sampler, batch_size=BATCH_SIZE)
+
+valid_data = TensorDataset(val_inputs, val_masks, val_tags)
+valid_sampler = SequentialSampler(valid_data)
+valid_dataloader = DataLoader(valid_data, sampler=valid_sampler, batch_size=BATCH_SIZE)
+
+print('Loaded training and validation data into DataLoaders successfully.')
+
+###
+
+# Initialize model
+model = BertForTokenClassification.from_pretrained('bert-base-cased',
+                                                   num_labels=NUM_LABELS)
+
+model.to(device)
+
+print(f'Initialized model and moved it to {device}.')
+
+###
+
+# Set hyperparameters (optimizer, weight decay, learning rate)
+
+if FULL_FINETUNING:
+    param_optimizer = list(model.named_parameters())
+    no_decay = ['bias', 'gamma', 'beta']
+    optimizer_grouped_parameters = [
+        {'params': [p for n, p in param_optimizer
+                    if not any(nd in n for nd in no_decay)],
+         'weight_decay_rate': 0.01},
+        {'params': [p for n, p in param_optimizer
+                    if any(nd in n for nd in no_decay)],
+         'weight_decay_rate': 0.0}
+    ]
+else:
+    param_optimizer = list(model.classifier.named_parameters())
+    optimizer_grouped_parameters = [{"params": [p for n, p in param_optimizer]}]
+
+optimizer = Adam(optimizer_grouped_parameters, lr=3e-5)
+
+print('Initialized optimizer and set hyperparameters.')
+
+###
+
+for _ in trange(EPOCHS, desc="Epoch"):
+
+    # Training loop
+    print('Starting training loop.')
+    model.train()
+    tr_loss = 0
+    nb_tr_examples, nb_tr_steps = 0, 0
+
+    for step, batch in enumerate(train_dataloader):
+
+        print('Starting a new training batch.')
+
+        # add batch to gpu
+        batch = tuple(t.to(device) for t in batch)
+        b_input_ids, b_input_mask, b_labels = batch
+
+        # forward pass
+        loss = model(b_input_ids, token_type_ids=None,
```

Comment:
this needs to be "loss, something_else" because of an update to the output of model()